src/main/java/com/amazonaws/encryptionsdk/keyrings/AwsKmsSymmetricKeyring.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    private EncryptionMaterials generateDataKey(final EncryptionMaterials encryptionMaterials) {
+        final DataKeyEncryptionDao.GenerateDataKeyResult result = this.dataKeyEncryptionDao.generateDataKey(
+            this.keyName, encryptionMaterials.getAlgorithm(), encryptionMaterials.getEncryptionContext());
+
+        return encryptionMaterials
+            .withCleartextDataKey(
+                result.getPlaintextDataKey(),
+                new KeyringTraceEntry(
+                    AWS_KMS_PROVIDER_ID,
+                    this.keyName.toString(),
+                    KeyringTraceFlag.GENERATED_DATA_KEY))
+            .withEncryptedDataKey(
+                new KeyBlob(result.getEncryptedDataKey()),
+                new KeyringTraceEntry(
+                    AWS_KMS_PROVIDER_ID,
+                    this.keyName.toString(),
+                    KeyringTraceFlag.ENCRYPTED_DATA_KEY,
+                    KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    private EncryptionMaterials encryptDataKey(final EncryptionMaterials encryptionMaterials) {
+        final EncryptedDataKey encryptedDataKey = this.dataKeyEncryptionDao.encryptDataKey(
+            this.keyName, encryptionMaterials.getCleartextDataKey(), encryptionMaterials.getEncryptionContext());
+
+        return encryptionMaterials.withEncryptedDataKey(
+            new KeyBlob(encryptedDataKey),
+            new KeyringTraceEntry(
+                AWS_KMS_PROVIDER_ID,
+                this.keyName.toString(),
+                KeyringTraceFlag.ENCRYPTED_DATA_KEY,
+                KeyringTraceFlag.SIGNED_ENCRYPTION_CONTEXT));
+    }
+
+    @Override
+    public DecryptionMaterials onDecrypt(DecryptionMaterials decryptionMaterials, List<? extends EncryptedDataKey> encryptedDataKeys) {
+        requireNonNull(decryptionMaterials, "decryptionMaterials are required");
+        requireNonNull(encryptedDataKeys, "encryptedDataKeys are required");
+
+        if (decryptionMaterials.hasCleartextDataKey() || encryptedDataKeys.isEmpty()) {
+            return decryptionMaterials;
+        }
+
+        for (EncryptedDataKey encryptedDataKey : encryptedDataKeys) {
+            if (okToDecrypt(encryptedDataKey)) {
+                try {
+                    final DataKeyEncryptionDao.DecryptDataKeyResult result = this.dataKeyEncryptionDao.decryptDataKey(
+                        encryptedDataKey, decryptionMaterials.getAlgorithm(), decryptionMaterials.getEncryptionContext());
+
+                    return decryptionMaterials.withCleartextDataKey(result.getPlaintextDataKey(),
+                        new KeyringTraceEntry(
+                            AWS_KMS_PROVIDER_ID,
+                            result.getKeyArn(),
+                            KeyringTraceFlag.DECRYPTED_DATA_KEY,
+                            KeyringTraceFlag.VERIFIED_ENCRYPTION_CONTEXT));
+                } catch (CannotUnwrapDataKeyException e) {
+                    continue;
+                }
```

Comment:
Good catch! We don't want to throw for discovery (since we can fail multiple keys in the same region). But we DO want to fail here - since there's a single key name associated with each keyring. Unless we get a list of EDKs all associated with that same key name, there shouldn't ever be a reason to try more than once. Updating.